cmd/snap-confine/snap-confine.c

```diff
@@ -409,18 +415,47 @@ int main(int argc, char **argv)
 	// For classic and confined snaps
 	sc_selinux_set_snap_execcon();
 #endif
-	if (sc_apply_seccomp_profile_for_security_tag(invocation.security_tag)) {
-		/* If the process is not explicitly unconfined then load the global
-		 * profile as well. */
-		sc_apply_global_seccomp_profile();
-	}
 	if (snap_context != NULL) {
 		setenv("SNAP_COOKIE", snap_context, 1);
 		// for compatibility, if facing older snapd.
 		setenv("SNAP_CONTEXT", snap_context, 1);
 	}
+
+	// Normally setuid/setgid not only permanently drops the UID/GID, but
+	// also clears the capabilities bounding sets (see "Effect of user ID
+	// changes on capabilities" in 'man capabilities'). To load a seccomp
+	// profile, we need either CAP_SYS_ADMIN or PR_SET_NO_NEW_PRIVS. Since
+	// NNP causes issues with AppArmor and exec transitions in certain
+	// snapd interfaces, keep CAP_SYS_ADMIN temporarily when we are
+	// permanently dropping privileges.
+	if (getresuid(&real_uid, &effective_uid, &saved_uid) != 0) {
+		die("getresuid failed");
+	}
+	debug("ruid: %d, euid: %d, suid: %d",
+	      real_uid, effective_uid, saved_uid);
+	struct __user_cap_header_struct hdr = { _LINUX_CAPABILITY_VERSION_3, 0 };
+	struct __user_cap_data_struct data[2] = { {0} };
+
+	// We only adjust the capabilities bounding set when we are permanently
+	// dropping.
```

Comment:
this comment seems the reverse or unclear:   we drop only if effective_uid == 0  but then keep_sys_admin is false because then effective_uid != 0 is false

Reply:
@pedronis 

Previously we would load the seccomp policy and if non-root permanently drop. This act of permanently dropping drops all capabilities and if the process was root, we would not drop, thus retaining all capabilities (letting apparmor mediate them).

The idea is that:
* because of the shuffle to drop before the seccomp policy load, we need to retain CAP_SYS_ADMIN to load the seccomp policy into the kernel
* if the snap command is running as non-root, we want to drop CAP_SYS_ADMIN after seccomp policy load to preserve previous behavior and not leak CAP_SYS_ADMIN to the non-root snap
* if the snap command is running as root (via sudo, daemon: simple, whatever) we do not want to drop CAP_SYS_ADMIN to preserve previous behavior (apparmor will continue to mediate its use)

In this code, the `keep_sys_admin` is the variable name to denote the non-root case where we want to permanently drop privileges but retain CAP_SYS_ADMIN until after we load the seccomp policy, when we drop CAP_SYS_ADMIN. Therefore, `effective_uid != 0 && saved_uid == 0` is the correct check for this moment in time (see a few lines above where we dropped for setup_user_data()) since only non-root will have effective_uid != 0 with saved_uid == 0. The comment is trying to convey this situation that we only need to keep CAP_SYS_ADMIN if we are non-root and need to permanently drop.

Perhaps the comment or the variable name need adjusting? (or both?)

Reply:
Yes, something needs to change, maybe what is confusing is the subsequent code:

	// Permanently drop if not root
	if (effective_uid == 0) {

either in the comment or the condition